internal/policy/container/has_modified_files.go

```diff
@@ -24,140 +28,158 @@ var _ check.Check = &HasModifiedFilesCheck{}
+
+		layerRefs[layerID] = packageFilesRef{
+			LayerFiles:        files,
+			LayerPackages:     pkgNameList,
+			LayerPackageFiles: packageFiles,
+			HasRPMDB:          true,
+		}
 	}
 
-	return files, dropped
+	return layerIDs, layerRefs, nil
 }
 
 // validate compares the list of LayerFiles and PackageFiles to see what PackageFiles
 // have been modified within the additional layers.
-//
-//nolint:unparam // ctx is unused. Keep for future use.
-func (p *HasModifiedFilesCheck) validate(ctx context.Context, packageFilesRef *packageFilesRef) (bool, error) {
+func (p *HasModifiedFilesCheck) validate(ctx context.Context, layerIDs []string, packageFiles map[string]packageFilesRef) (bool, error) {
 	logger := logr.FromContextOrDiscard(ctx)
 
-	layerFiles := packageFilesRef.LayerFiles
-	packageFiles := packageFilesRef.PackageFiles
-
-	// Determine the list of files that were a part of the base layer.
-	baseLayer := make(map[string]struct{}, len(layerFiles[0]))
-	for _, filename := range layerFiles[0] {
-		if _, ok := packageFiles[filename]; ok {
-			baseLayer[filename] = struct{}{}
-		}
-	}
-	layers := layerFiles[1:]
-
-	// Look for modifications in subsequent layers by determining
-	// if a file in a base layer exists in a subsequent layer.
-	modifiedFilesDetected := false
-	for _, layer := range layers {
-		for _, file := range layer {
-			if _, ok := baseLayer[file]; ok {
-				// This means the files exists in the base layer. This is a fail.
-				logger.V(log.DBG).Info("modified file detected", "filename", file)
-				modifiedFilesDetected = true
+	disallowedModifications := map[string]string{}
+	for idx, layerID := range layerIDs {
+		ref := packageFiles[layerID]
+		// if ref.HasRPMDB {
+		// 	// TODO: bpc
+		// 	// Something else needs to be other than just skipping because we have an RPM db in this layer.
+		// 	// It works for the *first* layer with an RPM db, since everything would be flagged, but we need to
+		// 	// figure out WHY a new RPM db is present.
+		// 	continue
+		// }
+		var modFound bool
+		for _, modifiedFile := range ref.LayerFiles {
+			// If it's a modifiedFile but this layer has an RPM db, that's allowed, but only if the
+			// package itself is updated.
+			if idx == 0 && ref.HasRPMDB {
+				// Just skip these in the first layer
+				continue
+			}
+			if _, found := ref.LayerPackageFiles[modifiedFile]; found {
+				previousPackageVersion, prevFound := packageFiles[layerIDs[idx-1]].LayerPackages[ref.LayerPackageFiles[modifiedFile]]
+				if !prevFound && ref.HasRPMDB {
+					// This is a net-new package file. Pass it.
+					continue
+				}
+				currentPackageVersion := ref.LayerPackages[ref.LayerPackageFiles[modifiedFile]]
+				if previousPackageVersion != currentPackageVersion {
+					// This appears like an update. This is allowed.
+					continue
+				}
+				modFound = true
+				disallowedModifications[modifiedFile] = layerID
 			}
 		}
+		if modFound {
+			logger.V(log.DBG).Info("found disallowed modification in layer")
```

Comment:
Does logging the layer buy us anything here? Or are we logging the layer/modification/etc elsewhere?

Reply:
+1 might as well log the layer.

Reply:
Not really. I took this straight from what @komish PoC'd. If we want to keep it, I should probably add the actually disallowed modification.

Reply:
I'd have to look closer, but the PoC definitely spat out a list of things we considered to be disallowed after doing this evaluation, and the layer.

```json
13:54:21 ~/projects/hasModifiedFilesTest/hasmodfiles-run-quay_io_komish_deleteme_test_mod_files_0_0_1
 ⤿ cat disallowedmods.json 
{
    "usr/bin/sha256sum": "sha256:2997d243150350c3d23695cf50cbfcbfa291a18553b7eaff006799e4dc718acf"
}
```

Reply:
This should be the case again.